### ISSUE TITLE
tshark display filters by ndpi protocol name

### ISSUE THREAD

**Issue by scroopyn00pers** [ntop/nDPI#505]
Hi everyone,

forgive me if I am asking a duplicate question, but I am at a loss with regards to ndpi-based display filters in tshark.

according to the docs, in wireshark you can 
> Specify an interface name (for live capture) or a pcap file path (for reading packets from a pcap file). (found at https://github.com/ntop/nDPI/tree/dev/wireshark)

this, of course, being done through the ndpi interface. Once I specify my capture interface (say, en2 or wlan0), I can successfully use a display filter, such as:

ndpi.protocol.name == SSL.Github

which is great! Unfortunately, I  am unable to replicate this using tshark's -Y flag (as in

`tshark -i en2 -Y "ndpi.protocol.name == SSL.Github"`
or 
`tshark -i ndpi -i en2 -Y "ndpi.protocol.name == SSL.Github"`

Do I need to modify ndpi.lua  and/or ndpiReader to be compatible with tshark?  Thanks for your patience, and have a nice day!

**Comment by bbqaaq** [comment]
Hi! I would like to share my results after 2 days efforts!

My target is to use ndpi to enhance the bittorrent detection using tshark, however, there aren't any good instruction on how to combine this two things.

Following the instructions in https://sharkfesteurope.wireshark.org/assets/presentations17eu/19.pdf
it's not that hard to integrate ndpi into wireshark. With GUI wireshark the result is perfect.
But my big question is how to use ndpi with tshark.

At first I was like you, trying to use ndpi to capture eth0 (or en0) interface, with the argument like
`tshark -i ndpi` or `tshark -i ndpi -i eth0`
As the result, you will see I got nothing from that interface.
![fig1](https://user-images.githubusercontent.com/6056605/34813859-7d5c7148-f6e6-11e7-8d2c-80dde2d350da.png)

The problem is, I should have provided the interface option to the ndpi, not the tshark.
So after hours of searching and trying, I finally figure out that I need to change the wireshark preferences.
According to tshark doc (https://www.wireshark.org/docs/man-pages/tshark.html):

> -o <preference>:<value>
Set a preference value, overriding the default value and any value read from a preference file. The argument to the option is a string of the form prefname:value, where prefname is the name of the preference (which is the same name that would appear in the preference file), and value is the value to which it should be set.

So we need to find out what to set into the preference.
![fig2](https://user-images.githubusercontent.com/6056605/34813998-1bf0cf66-f6e7-11e7-80ec-7977f1df432b.png)
![fig3](https://user-images.githubusercontent.com/6056605/34813999-1c217eb8-f6e7-11e7-8f33-cf0ed085241f.png)

In wireshark I can see that `extcap.ndpi.i` and `extcap.ndpi.9` is changed when I changed the ndpi interface settings. But what do they mean? After reading the ndpiReader doc, indeed `extcap.ndpi.i` the settings for the interface/pcap for ndpi to reader, and `extcap.ndpi.9` is the target protocol capture by ndpi.

After a long time searching, the magic command is
`tshark -i ndpi -o extcap.ndpi.i:/what/ever/path/cap.pcap  -o extcap.ndpi.9:-1`

But unfortunately, I cannot use the -Y display filter in tshark. Seems it's working in wireshark, but no in tshark. But at least Im glad here to share to someone who get lost in using ndpi with tshark.

**Comment by bbqaaq** [comment]
Indeed the display filter should be related to the ndpi.lua script, not the extcap (ndpiReader) module.

Im sure that my lua script  is being executed (with the log), but I have no idea why the packet's protocol doesn't display what the wireshark show. (In wireshark it shows BitTorrent suggested by nDPI, but in tshark its still showing UDP sugggested by wireshark)

**Comment by kYroL01** [comment]
Hi all.
So, it seems to be an inconsistency between WIreshark and Tshark with nDPI filtering.
We're trying to figure out a solution asap.
Any other informations or suggestions will be appreciate.

**Comment by bbqaaq** [comment]
I suggest to add more instructions on how to use tshark with nDPI, in the readme file of wireshark/README.md

I used nearly 2 days to figure out how to make tshark work with it, which is a nasty work. At least someone who need to do similar job with me can have more information for them.

**Comment by kYroL01** [comment]
@bbqaaq I'm agree with you and thanks to evidence us this fact.

**Comment by jiangdu8888** [comment]
@bbqaaq hello how do you figure out it ,i can not find ndpi capture interface with wireshark

**Comment by bbqaaq** [comment]
@jiangdu8888 Sorry for late reply.

Indeed I was following this guide to try to setup the ndpi with wireshark.
https://sharkfesteurope.wireshark.org/assets/presentations17eu/19.pdf

If you still have problem with it, try to google it on how to setup lua/plugin with wireshark.
I think you can find some tutorial on how to set it up.

**Comment by Lizeth2989** [comment]
Hey guys,
I installed nDPI in wireshark using this guide https://sharkfesteurope.wireshark.org/assets/presentations17eu/19.pdf

First, I did it in Windows (wireshark 3.5) and the nDPI interface never appeared. Only Lua menu.

Later, I used an ubuntu (wireshark 2.6) and the nDPI interface appeared. However, when I click on it, gives me this error

"No packets capture, As no data was captured, closing the temporary capture file. Help ..."

Ofc I captured packets but I do not receive the wireshark interface options (ppt page 33). Someone can help me !

Thank you
